DOC: Exposed arguments in plot.kde

tommyod opened the pull request:
The documentation for `plot.kde` did not show the `bw_method` and `ind` arguments, which are specific to `plot.kde` (and `plot.density`, which refers to the same method).

There is also a change in the actual code, and a corresponding test added. I added an option for `ind` to be an integer number of sample points, instead of necessarily a `np.array`. I image that a user typically just wants a number of equidistant sample points, and does not want to construct an array.

- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Comments very welcome.

tommyod committed:
Exposed arguments in plot.kde, added number of sample points as option

tommyod committed:
Added a test for plot.kde with  as an integer

pep8speaks commented:
Hello @tommyod! Thanks for submitting the PR.

 - In the file [`pandas/plotting/_core.py`](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py), following are the PEP8 issues :

> [Line 2592:28](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2592): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2592:30](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2592): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2592:40](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2592): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2592:42](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2592): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2592:48](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2592): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 2606:75](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2606): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 2615:42](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2615): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2615:44](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2615): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2615:59](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2615): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2615:61](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2615): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2770:28](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2770): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2770:30](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2770): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2770:40](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2770): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2770:42](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2770): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2770:48](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2770): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','
> [Line 2784:75](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2784): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 2793:42](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2793): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2793:44](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2793): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2793:59](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2793): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2793:61](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2793): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 2880:1](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/plotting/_core.py#L2880): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

 - In the file [`pandas/tests/plotting/test_series.py`](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/tests/plotting/test_series.py), following are the PEP8 issues :

> [Line 901:1](https://github.com/pandas-dev/pandas/blob/8000f945de320a8e1c5f0e1f98042e3480cc9e3a/pandas/tests/plotting/test_series.py#L901): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

tommyod commented:
Ran the `flake8` tool while in the wrong subdirectory locally. Will fix the errors tomorrow and push.

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -1398,6 +1398,10 @@ def _get_ind(self, y):
             sample_range = np.nanmax(y) - np.nanmin(y)
             ind = np.linspace(np.nanmin(y) - 0.5 * sample_range,
                               np.nanmax(y) + 0.5 * sample_range, 1000)
+        elif isinstance(self.ind, (int, np.int)):
```
Use  `isinstance(self.ind, numbers.Integral)`. It should catch both of these (may have to import `numbers`).

jreback replied:
instead of this, just use ``is_integer``, which is our version of this and handles all the cases.
import from ``pandas.core.dtypes.common`` (if not already imported)

TomAugspurger commented on pandas/tests/plotting/test_series.py:
```diff
@@ -622,6 +622,7 @@ def test_kde_kwargs(self):
             pytest.skip("mpl is not supported")
 
         from numpy import linspace
+        _check_plot_works(self.ts.plot.kde, bw_method=None, ind=20)
```
Could you add a test with `np.int(20)` as well.

jreback replied:
can you add a use with ``bw_method`` as a string.

TomAugspurger reviewed (commented):
Thanks. Could fixup the linting errors, and add a release note in `whatsnew/v0.23.0.txt`?

tommyod committed:
Added whatsnew. Fixed flake8 errors. Used is_integer  to infer type.

tommyod commented:
@jreback, @TomAugspurger  - I have addressed the comments.

- Wrote an additional test.
- Fixed linting errors.
- Type checked with `is_integer`.
- Added `whatsnew` comment.

tommyod commented:
@jreback Friendly reminder to consider this PR, in case it has been forgotten. If there is anything else that needs to be done, please let me know.

TomAugspurger commented on pandas/plotting/_core.py:
```diff
@@ -2753,20 +2767,30 @@ def hist(self, by=None, bins=10, **kwds):
+            The method used to calculate the estimator bandwidth.  This can be
+            'scott', 'silverman', a scalar constant or a callable.
+            If None (default), 'scott' is used.
+            See :scipy:class:`stats.gaussian_kde` for more information.
```
Have you verified that this link works? I thought it would be

    :func:`scipy.stats.gaussian_kde`.

tommyod replied:
Agreed. Changed it.

tommyod committed:
Updated scipy reference

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -478,7 +478,7 @@ Plotting
 - :func: `DataFrame.plot` now raises a ``ValueError`` when the ``x`` or ``y`` argument is improperly formed (:issue:`18671`)
 - Bug in formatting tick labels with ``datetime.time()`` and fractional seconds (:issue:`18478`).
--
+- The arguments ``ind`` and ``bw_method`` are added to the docstring of :meth:`Series.plot.kde` (:issue:`18461`). The argument ``ind`` may now also be an integer (number of sample points).
```
``:meth:`Series.plot.kde` `` has exposed the args......in the doc-string.

jreback commented on pandas/tests/plotting/test_series.py:
```diff
@@ -622,6 +622,8 @@ def test_kde_kwargs(self):
             pytest.skip("mpl is not supported")
 
         from numpy import linspace
```
don't import here, rather use ``np.linspace``

jreback commented on pandas/tests/plotting/test_series.py:
```diff
@@ -622,6 +622,8 @@ def test_kde_kwargs(self):
 
         from numpy import linspace
+        _check_plot_works(self.ts.plot.kde, bw_method=None, ind=20)
+        _check_plot_works(self.ts.plot.kde, bw_method=None, ind=np.int(20))
```
ideally would parametrize this test instead of repeating.

jreback reviewed (changes requested):
small changes.

tommyod committed:
Added test, rewrote whatsnew, removed import

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2753,20 +2767,30 @@ def hist(self, by=None, bins=10, **kwds):
+            spaced points are used.
         `**kwds` : optional
-            Keyword arguments to pass on to :py:meth:`pandas.DataFrame.plot`.
+            Keyword arguments to pass on to :py:meth:`pandas.Series.plot`.
```
is this change right? This the DataFrame one

tommyod replied:
Good catch. My bad. It should be ``pandas.DataFrame.plot``, and I have made the change.

tommyod committed:
Changed from Series to DataFrame in doc

tommyod committed:
Fixed PEP8 errors in test file

tommyod committed:
Fixed typo which made tests crash

tommyod commented:
@jreback I believe this PR is good to go now.

TomAugspurger commented:
Thanks @tommyod!
